fili-core/src/test/groovy/com/yahoo/bard/webservice/async/AsyncFunctionalSpec.groovy

```diff
@@ -0,0 +1,211 @@
+     */
+    abstract Closure<String> getFakeDruidResponse()
+
+    /**
+     * Returns the status code that the fake Druid should send when sending results back.
+     *
+     * @return The status code that the fake Druid should use when sending results back, defaults to 200 (OK)
+     */
+    int getDruidStatusCode() {
+        return 200
+    }
+
+    def setup() {
+        // Create the test web container to test the resources
+        jtb = buildTestBinder()
+
+        populatePhysicalTableAvailability()
+    }
+
+    /**
+     * Populates the interval availability of the physical tables.
+     * <p>
+     * By default, every Physical table believes that it has complete data from January 1st 2010 to December 31 2500.
+     */
+    void populatePhysicalTableAvailability() {
+        AvailabilityTestingUtils.populatePhysicalTableCacheIntervals(jtb, new Interval("2010-01-01/2500-12-31"))
+    }
+
+    def cleanup() {
+        // Release the test web container
+        jtb.tearDown()
+    }
+
+    JerseyTestBinder buildTestBinder() {
+        new JerseyTestBinder(getResourceClasses())
+    }
+
+    void validateJson(String json) {
+        MAPPER.readTree(json)
+    }
+
+    @Timeout(10)
+    def "The asynchronous workflow executes correctly"() {
+        given: "Initialize the backend with the fake Druid response"
+        injectDruidResponse(getFakeDruidResponse())
+
+        and: "The map of all previous responses"
+        Map<String, Response> previousResponses = [:]
+
+        and: "The closures that perform verification against the results"
+        Map<String, Closure<Void>> resultValidations = getResultAssertions()
+
+        and: "The query parameters for each request"
+        Map<String, Closure<Map<String, List<String>>>> queryParameters = getQueryParameters()
+
+        expect: "All the asynchronous interactions behave appropriately"
+        getResultsToTargetFunctions().each {interactionName, resultsToTarget ->
+            Response response = makeAbstractRequest(
+                    resultsToTarget(previousResponses),
+                    queryParameters.get(interactionName)(previousResponses)
+            )
+            response.bufferEntity()
+            resultValidations[interactionName](response)
+            previousResponses.put(interactionName, response)
+        }
+    }
+
+    /**
+     *  Injects the closure generating the fake Druid response into the Druid backend used by the test harness, if
+     *  applicable.
+     *
+     *  @param  druidResponse The fake response to be injected.
+     */
+    def injectDruidResponse(Closure<String> druidResponse) {
+        if (jtb.nonUiDruidWebService instanceof TestDruidWebService) {
+            int statusCode = getDruidStatusCode()
+            jtb.nonUiDruidWebService.statusCode = statusCode
+            if (statusCode == 200) {
+                jtb.nonUiDruidWebService.jsonResponse = druidResponse
```

Comment:
No. The `jsonResponse` in the DruidWebService is now a `Function`, not a `String` in order to support delaying the response.